src/newton_solver.cpp

```diff
@@ -105,144 +78,71 @@ void NewtonSolver::computeNewtonSensis(
     }
 }
 
-NewtonSolverDense::NewtonSolverDense(Model const& model)
-    : NewtonSolver(model)
-    , Jtmp_(model.nx_solver, model.nx_solver)
-    , linsol_(SUNLinSol_Dense(x_.getNVector(), Jtmp_)) {
-    auto status = SUNLinSolInitialize_Dense(linsol_);
-    if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolInitialize_Dense");
-}
-
-void NewtonSolverDense::prepareLinearSystem(
+void NewtonSolver::prepareLinearSystem(
     Model& model, SimulationState const& state
 ) {
-    model.fJ(state.t, 0.0, state.x, state.dx, xdot_, Jtmp_);
-    Jtmp_.refresh();
-    auto status = SUNLinSolSetup_Dense(linsol_, Jtmp_);
-    if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSetup_Dense");
+    auto& J = linsol_->getMatrix();
+    if (J.matrix_id() == SUNMATRIX_SPARSE) {
+        model.fJSparse(state.t, 0.0, state.x, state.dx, xdot_, J);
+    } else {
+        model.fJ(state.t, 0.0, state.x, state.dx, xdot_, J);
+    }
+    J.refresh();
+    try {
+        linsol_->setup();
+    } catch (AmiException const& e) {
+        throw NewtonFailure(AMICI_SINGULAR_JACOBIAN, e.what());
+    }
 }
 
-void NewtonSolverDense::prepareLinearSystemB(
+void NewtonSolver::prepareLinearSystemB(
     Model& model, SimulationState const& state
 ) {
-    model.fJB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, Jtmp_);
-    Jtmp_.refresh();
-    auto status = SUNLinSolSetup_Dense(linsol_, Jtmp_);
-    if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSetup_Dense");
+    auto& J = linsol_->getMatrix();
+    if (J.matrix_id() == SUNMATRIX_SPARSE) {
+        model.fJSparseB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, J);
+    } else {
+        model.fJB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, J);
+    }
+    J.refresh();
+    try {
+        linsol_->setup();
+    } catch (AmiException const& e) {
+        throw NewtonFailure(AMICI_SINGULAR_JACOBIAN, e.what());
+    }
 }
 
-void NewtonSolverDense::solveLinearSystem(AmiVector& rhs) {
-    auto status = SUNLinSolSolve_Dense(
-        linsol_, Jtmp_, rhs.getNVector(), rhs.getNVector(), 0.0
-    );
-    Jtmp_.refresh();
+void NewtonSolver::solveLinearSystem(AmiVector& rhs) {
     // last argument is tolerance and does not have any influence on result
-
+    auto status = linsol_->solve(rhs.getNVector(), rhs.getNVector(), 0.0);
     if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSolve_Dense");
+        throw NewtonFailure(status, "SUNLinSolSolve");
 }
 
-void NewtonSolverDense::reinitialize() {
-    /* dense solver does not need reinitialization */
+void NewtonSolver::reinitialize() {
+    // dense solver does not need reinitialization
```

Comment:
```suggestion
    // direct solvers does not need reinitialization
```
what about iterative solvers?

Reply:
They are already filtered out in the constructor: https://github.com/AMICI-dev/AMICI/blob/12ee3ebbcb35e7536cd4d5c3b6447efc93a79d9b/src/newton_solver.cpp#L19-L30

Some methods will have to be updated when more solvers are to be supported.

Reply:
Added comment+exception regarding the need for updating those two functions to support additional solvers.